[FLINK-19145][walkthroughs] Add PyFlink-walkthrough to Flink playground. (#16)
apache/flink-playgrounds: pyflink-walkthrough/README.md

# pyflink-walkthrough

## Background

> In this playground, you will learn how to manage and run PyFlink Jobs. The pipeline of this walkthrough reads data from Kafka, performs aggregations and writes results to Elasticsearch visualized via Kibana. The environment is managed by Docker so that all you need is a docker on your computer.

Reviewer: ```suggestion
In this playground, you will learn how to build and run an end-to-end PyFlink pipeline for data analytics, covering the following steps:

 * Reading data from a Kafka source;
 * Creating and using a [UDF](https://ci.apache.org/projects/flink/flink-docs-release-1.11/dev/python/table-api-users-guide/udfs/python_udfs.html);
 * Performing a simple aggregation over the source data;
 * Writing the results to Elasticsearch and visualizing them in Kibana. 
 
 The environment is based on Docker Compose, so the only requirement is that you have [Docker](https://docs.docker.com/get-docker/) installed in your machine.
```

- Kafka

> Kafka is used to store input data in this walkthrough. The script [generate_source_data.py](https://github.com/hequn8128/pyflink-walkthrough/blob/master/generate_source_data.py) is used to generate transaction data and writes into the payment_msg kafka topic. Each record includes 5 fields: 
> ```text
> {"createTime": "2020-08-12 06:29:02", "orderId": 1597213797, "payAmount": 28306.44976403719, "payPlatform": 0, "provinceId": 4}
> ```
> ```text
> createTime: The creation time of the transaction. 
> orderId: The id of the current transaction.
> payAmount: The pay amount of the current transaction.
> payPlatform: The platform used to pay the order, pc or mobile.
> provinceId: The id of the province for the user. 
> ```
> 
> - Generator 
> 
> A simple data generator is provided that continuously writes new records into Kafka. 
> You can use the following command to read data in kafka and check whether the data is generated correctly.

Reviewer: ```suggestion
You will be using Kafka to store sample input data about payment transactions. A simple data generator `[generate_source_data.py](https://github.com/hequn8128/pyflink-walkthrough/blob/master/generate_source_data.py)` is provided to continuously write new records to the `payment_msg` Kafka topic. Each record is structured as follows: 

`{"createTime": "2020-08-12 06:29:02", "orderId": 1597213797, "payAmount": 28306.44976403719, "payPlatform": 0, "provinceId": 4}`

* `createTime`: The creation time of the transaction. 
* `orderId`: The ID of the current transaction.
* `payAmount`: The pay amount of the current transaction.
* `payPlatform`: The platform used to pay the order (pc or mobile).
* `provinceId`: The ID of the province the user is located in. 

You can use the following command to read data from the Kafka topic and check whether it's generated correctly:
```

```shell script
$ docker-compose exec kafka kafka-console-consumer.sh --bootstrap-server kafka:9092 --topic payment_msg
{"createTime":"2020-07-27 09:25:32.77","orderId":1595841867217,"payAmount":7732.44,"payPlatform":0,"provinceId":3}
{"createTime":"2020-07-27 09:25:33.231","orderId":1595841867218,"payAmount":75774.05,"payPlatform":0,"provinceId":3}
{"createTime":"2020-07-27 09:25:33.72","orderId":1595841867219,"payAmount":65908.55,"payPlatform":0,"provinceId":0}
{"createTime":"2020-07-27 09:25:34.216","orderId":1595841867220,"payAmount":15341.11,"payPlatform":0,"provinceId":1}
{"createTime":"2020-07-27 09:25:34.698","orderId":1595841867221,"payAmount":37504.42,"payPlatform":0,"provinceId":0}
```

- PyFlink

> The transaction data is processed by a PyFlink job, [payment_msg_proccessing.py](https://github.com/hequn8128/pyflink-walkthrough/blob/master/payment_msg_proccessing.py). The job maps the province id to province name for better demonstration using a Python UDF and then sums the payment for each province using a group aggregate. 

Reviewer: ```suggestion
The transaction data will be processed with PyFlink using the Python script `[payment_msg_proccessing.py](https://github.com/hequn8128/pyflink-walkthrough/blob/master/payment_msg_proccessing.py)`. This script will first map the `provinceId` in the input records to its corresponding province name using a Python UDF, and then sum the transaction amount for each province using a group aggregate. 
```

- ElasticSearch

ElasticSearch is used to store upstream processing results and provide efficient query service.

- Kibana

> Kibana is an open source data visualization dashboard for ElasticSearch. We use it to visualize our processing results.

Reviewer: ```suggestion
Kibana is an open source data visualization dashboard for ElasticSearch. You will use it to visualize the results of your PyFlink pipeline.
```

## Setup

> The pyflink-walkthrough requires a custom Docker image, as well as public images for Flink, Elasticsearch, Kafka, and ZooKeeper. 
> 
> The [docker-compose.yaml](https://github.com/hequn8128/pyflink-walkthrough/blob/master/docker-compose.yml) file of the pyflink-walkthrough is located in the `pyflink-walkthrough` root directory.

Reviewer: ```suggestion
As mentioned, the environment for this walkthrough is based on Docker Compose; and uses a custom image to spin up Flink (JobManager+TaskManager), Kafka+Zookeeper, the data generator and Elasticsearch+Kibana containers.

You can find the [docker-compose.yaml](https://github.com/hequn8128/pyflink-walkthrough/blob/master/docker-compose.yml) file in the `pyflink-walkthrough` root directory.
```

> ### Building the custom Docker image

Reviewer: ```suggestion
### Building the Docker image
```

> Build the Docker image by running

Reviewer: ```suggestion
First, build the Docker image by running:
```

```bash
docker-compose build
```

### Starting the Playground

> Once you built the Docker image, run the following command to start the playground

Reviewer: ```suggestion
Once the Docker image build is complete, run the following command to start the playground:
```

```bash
docker-compose up -d
```

> You can check if the playground was successfully started by accessing the WebUI of(You may need to wait about 1 min before all services come up.):

Reviewer: ```suggestion
One way of checking if the playground was successfully started is accessing some of the services exposed:
```

1. visiting Flink Web UI [http://localhost:8081](http://localhost:8081).
2. visiting Elasticsearch [http://localhost:9200](http://localhost:9200).
3. visiting Kibana [http://localhost:5601](http://localhost:5601).

> 

Reviewer: ```suggestion
**Note:** you may need to wait around 1 minute before all the services come up.

```

### Stopping the Playground

> To stop the playground, run the following command

Reviewer: ```suggestion
To stop the playground, run the following command:
```

```bash
docker-compose down
```

> ## Run jobs

Reviewer: ```suggestion
## Running the PyFlink Job
```

1. Submit the PyFlink job.
```shell script
$ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
```

> 2. Open [kibana ui](http://localhost:5601) and choose the dashboard: payment_dashboard

Reviewer: ```suggestion
2. Navigate to the [Kibana UI](http://localhost:5601) and choose the pre-created dashboard `payment_dashboard`.
```

![image](pic/dash_board.png)

![image](pic/final.png)

> 3. Stop PyFlink job:

Reviewer: ```suggestion
3. Stop the PyFlink job:
```

> Visit [http://localhost:8081/#/overview](http://localhost:8081/#/overview) , select the job and click `Cancle`.

Reviewer: ```suggestion
Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview), select the job and click `Cancel` on the upper right side.
```

# pyflink-walkthrough

## Background

In this playground, you will learn how to build and run an end-to-end PyFlink pipeline for data analytics, covering the following steps:

* Reading data from a Kafka source;
* Creating data using a [UDF](https://ci.apache.org/projects/flink/flink-docs-release-1.11/dev/python/table-api-users-guide/udfs/python_udfs.html);
* Performing a simple aggregation over the source data;
* Writing the results to Elasticsearch and visualizing them in Kibana.

### Kafka
You will be using Kafka to store sample input data about payment transactions. A simple data generator [generate_source_data.py](generator/generate_source_data.py) is provided to
continuously write new records to the `payment_msg` Kafka topic. Each record is structured as follows:
 
`{"createTime": "2020-08-12 06:29:02", "orderId": 1597213797, "payAmount": 28306.44976403719, "payPlatform": 0, "provinceId": 4}`

* `createTime`: The creation time of the transaction. 
* `orderId`: The id of the current transaction.

> * `payAmount`: The pay amount of the current transaction.

Reviewer 2: ```suggestion
* `payAmount`: The amount being paid with this transaction.
```

> * `payPlatform`: The platform used to pay the order, pc or mobile.

Reviewer 2: ```suggestion
* `payPlatform`: The platform used to create this payment: pc or mobile.
```

* `provinceId`: The id of the province for the user. 

You can use the following command to read data from the Kafka topic and check whether it's generated correctly:
```shell script
$ docker-compose exec kafka kafka-console-consumer.sh --bootstrap-server kafka:9092 --topic payment_msg
{"createTime":"2020-07-27 09:25:32.77","orderId":1595841867217,"payAmount":7732.44,"payPlatform":0,"provinceId":3}
{"createTime":"2020-07-27 09:25:33.231","orderId":1595841867218,"payAmount":75774.05,"payPlatform":0,"provinceId":3}
{"createTime":"2020-07-27 09:25:33.72","orderId":1595841867219,"payAmount":65908.55,"payPlatform":0,"provinceId":0}
{"createTime":"2020-07-27 09:25:34.216","orderId":1595841867220,"payAmount":15341.11,"payPlatform":0,"provinceId":1}
{"createTime":"2020-07-27 09:25:34.698","orderId":1595841867221,"payAmount":37504.42,"payPlatform":0,"provinceId":0}
```

### PyFlink

The transaction data will be processed with PyFlink using the Python script [payment_msg_processing.py](payment_msg_proccessing.py).
This script will first map the `provinceId` in the input records to its corresponding province name

> using a Python UDF, and them sum the transaction amount for each province using a group aggregate. 

Reviewer 2: ```suggestion
using a Python UDF, and then compute the sum of the transaction amounts for each province. 
```

### ElasticSearch

> ElasticSearch is used to store upstream processing results and provide efficient query service.

Reviewer 2: ```suggestion
ElasticSearch is used to store the results and to provide an efficient query service.
```

### Kibana

Kibana is an open source data visualization dashboard for ElasticSearch. You will use it to visualize 
the results of your PyFlink pipeline.

## Setup

As mentioned, the environment for this walkthrough is based on Docker Compose; and uses a custom image
to spin up Flink (JobManager+TaskManager), Kafka+Zookeeper, the data generator and Elasticsearch+Kibana

> containers.

Reviewer 2: ```suggestion
As mentioned, the environment for this walkthrough is based on Docker Compose. It uses a custom image
to spin up Flink (JobManager+TaskManager), Kafka+Zookeeper, the data generator, and Elasticsearch+Kibana
containers.
```

Your can find the [docker-compose.yaml](docker-compose.yml) file of the pyflink-walkthrough is located in the `pyflink-walkthrough` root directory.

### Building the Docker image

First, build the Docker image by running:

```bash
docker-compose build
```

### Starting the Playground

Once the Docker image build is complete, run the following command to start the playground:

```bash
docker-compose up -d
```

> One way of checking if the playground was successfully started is accessing some of the services exposed:

Reviewer 2: ```suggestion
One way of checking if the playground was successfully started is to access some of the services that are exposed:
```

1. visiting Flink Web UI [http://localhost:8081](http://localhost:8081).
2. visiting Elasticsearch [http://localhost:9200](http://localhost:9200).
3. visiting Kibana [http://localhost:5601](http://localhost:5601).

**Note:** you may need to wait around 1 minute before all the services come up.

### Stopping the Playground

To stop the playground, run the following command:

```bash
docker-compose down
```

> ## Runing the PyFlink job

Reviewer 2: ```suggestion
## Running the PyFlink job
```

1. Submit the PyFlink job.
```shell script
$ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
```

2. Navigate to the [Kibana UI](http://localhost:5601) and choose the pre-created dashboard `payment_dashboard`.

![image](pic/dash_board.png)

![image](pic/final.png)

3. Stop the PyFlink job:

> Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview) , select the job and click `Cancle` on the upper right side.

Reviewer 2: ```suggestion
Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview) , select the job and click `Cancel` on the upper right side.
```

# pyflink-walkthrough

## Background

In this playground, you will learn how to build and run an end-to-end PyFlink pipeline for data analytics, covering the following steps:

* Reading data from a Kafka source;
* Creating data using a [UDF](https://ci.apache.org/projects/flink/flink-docs-release-1.11/dev/python/table-api-users-guide/udfs/python_udfs.html);
* Performing a simple aggregation over the source data;
* Writing the results to Elasticsearch and visualizing them in Kibana.

The environment is based on Docker Compose, so the only requirement is that you have [Docker](https://docs.docker.com/get-docker/) 
installed in your machine.

### Kafka

> You will be using Kafka to store sample input data about payment transactions. A simple data generator [generate_source_data.py](generator/generate_source_data.py) is provided to

Author: Here I would like to use `H3` header for each component (Kafka, PyFlink, Elaticsearch, Kibana) introduction to make them more recognizable.

* `payPlatform`: The platform used to create this payment: pc or mobile.
* `provinceId`: The id of the province for the user. 

You can use the following command to read data from the Kafka topic and check whether it's generated correctly:
```shell script
$ docker-compose exec kafka kafka-console-consumer.sh --bootstrap-server kafka:9092 --topic payment_msg
{"createTime":"2020-07-27 09:25:32.77","orderId":1595841867217,"payAmount":7732.44,"payPlatform":0,"provinceId":3}
{"createTime":"2020-07-27 09:25:33.231","orderId":1595841867218,"payAmount":75774.05,"payPlatform":0,"provinceId":3}
{"createTime":"2020-07-27 09:25:33.72","orderId":1595841867219,"payAmount":65908.55,"payPlatform":0,"provinceId":0}
{"createTime":"2020-07-27 09:25:34.216","orderId":1595841867220,"payAmount":15341.11,"payPlatform":0,"provinceId":1}
{"createTime":"2020-07-27 09:25:34.698","orderId":1595841867221,"payAmount":37504.42,"payPlatform":0,"provinceId":0}
```
You can also create a new topic by executing the following command:
```shell script
$ docker-compose exec kafka kafka-topics.sh --bootstrap-server kafka:9092 --create --topic <YOUR-TOPIC-NAME> --partitions 8 --replication-factor 1

> ```

Reviewer 2: It is confusing to present, this early in the documentation, details of how to work with Kafka that won't actually work before having built and launched the docker image.  

Some, or perhaps all, of the information in this background section needs to come later, in a "just in time" fashion, when it is needed. Readers want to make a single, linear pass through one of these walkthroughs, executing commands as they go. Presenting shell commands before they can be used will confuse some readers.

Author: I will move this part to the later section between before starting the PyFlink job after docker compose up.

### PyFlink

The transaction data will be processed with PyFlink using the Python script [payment_msg_processing.py](payment_msg_proccessing.py).
This script will first map the `provinceId` in the input records to its corresponding province name using a Python UDF, 
and then compute the sum of the transaction amounts for each province.

### ElasticSearch

ElasticSearch is used to store the results and to provide an efficient query service.

### Kibana

Kibana is an open source data visualization dashboard for ElasticSearch. You will use it to visualize 
the results of your PyFlink pipeline.

## Setup

As mentioned, the environment for this walkthrough is based on Docker Compose; It uses a custom image
to spin up Flink (JobManager+TaskManager), Kafka+Zookeeper, the data generator, and Elasticsearch+Kibana containers.

You can find the [docker-compose.yaml](docker-compose.yml) file of the pyflink-walkthrough in the `pyflink-walkthrough` root directory.

### Building the Docker image

First, build the Docker image by running:

```bash

> docker-compose build

Reviewer 2: When I tried this, it took me a couple of minutes to realize it hadn't done the right things because I was still in the parent directory.

```suggestion
cd pyflink-walkthrough
docker-compose build
```

```

### Starting the Playground

Once the Docker image build is complete, run the following command to start the playground:

```bash
docker-compose up -d
```

One way of checking if the playground was successfully started is to access some of the services that are exposed:

1. visiting Flink Web UI [http://localhost:8081](http://localhost:8081).

> 2. visiting Elasticsearch [http://localhost:9200](http://localhost:9200).

Reviewer 2: This is what I see at localhost:9200

![image](https://user-images.githubusercontent.com/43608/94797024-d8833580-03df-11eb-8cef-1928fdb42695.png)

Author: This simply indicates that the Elasticsearch service is successfully up  when you visit this link.

3. visiting Kibana [http://localhost:5601](http://localhost:5601).

**Note:** you may need to wait around 1 minute before all the services come up.

### Stopping the Playground

To stop the playground, run the following command:

```bash
docker-compose down

> ```

Reviewer 2: This should come later. If it is left here, some folks will blindly stop the playground at this point.

Author: Yes, it should be placed in the end of the walkthrough.

## Running the PyFlink job

1. Submit the PyFlink job.
```shell script

> $ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d

Reviewer 2: I suggest being consistent, and putting a $ prompt in front of all of the other shell commands in this README as well.

```
Navigate to the [Flink Web UI](http://localhost:8081) after the job is submitted successfully. There should be a job in the running job list.
Click the job to get more details. You should see that the `StreamGraph` of the `payment_msg_proccessing` consists of two nodes, each with a parallelism of 1. 

> There is also a table in the bottom of the page that shows the metrics of every node (e.g. bytes received/sent, records received/sent).

Reviewer 2: ```suggestion
There is also a table in the bottom of the page that shows some metrics for each node (e.g. bytes received/sent, records received/sent). Note that Flink's metrics only report bytes and records communicated within the Flink cluster, and so will always report 0 bytes and 0 records received by sources, and 0 bytes and 0 records sent to sinks — so don't be confused that nothing is reported as being read from Kafka, or written to Elasticsearch.
```

![image](pic/submitted.png)

![image](pic/detail.png)    

> 2. Navigate to the [Kibana UI](http://localhost:5601) and choose the pre-created dashboard `payment_dashboard`. There will be a vertical bar chart and a pie chart demonstrating the total amount and the proportion of each province.

Reviewer 2: This didn't work for me. localhost:5061 redirected to http://localhost:5601/app/kibana#/home where I see this:

![image](https://user-images.githubusercontent.com/43608/94797899-1c2a6f00-03e1-11eb-9868-7a23daac95f8.png)

Author: Sorry, the localhost:5061 would lead you to the home page of Kibana, then you need to open the menu list by clicking the menu button in the upper left corner, then choose the  `Dashboard` item to turn the dashboard page. I will make it more clear.

![image](pic/dash_board.png)

![image](pic/chart.png)

3. Stop the PyFlink job:

> Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview) , select the job and click `Cancel` on the upper right side.

Reviewer 2: ```suggestion
Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview), select the job, and click `Cancel Job` in the upper right corner.
```

![image](pic/cancel.png)

## Further Explorations

If you would like to explore this example more deeply, you can edit [payment_msg_processing.py](payment_msg_proccessing.py)
or create new PyFlink projects that perform more complex processing. You can do this locally under
the `pyflink-walkthrough` directory, since it is mounted on the `jobmanager` docker container.

Ideas:
* Add your own Kafka source table;
* Create a new index for the Elasticsearch sink;
* Count the number of transactions, grouped by a 1 minute [tumbling window](https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/tableApi.html#tumble-tumbling-windows) and `payPlatform`.

After making a modification, you can submit the new job by executing the same command mentioned at 
[Running the PyFlink Job](#running-the-pyflink-job)
```shell script
$ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
```

Furthermore, you can also [create new Kibana dashboards](https://www.elastic.co/guide/en/kibana/7.8/dashboard-create-new-dashboard.html) 

> to visualize other aspects of the data in the Elasticsearch.

Reviewer 2: ```suggestion
to visualize other aspects of the data in Elasticsearch.
```

```
You can also create a new topic by executing the following command:
```shell script
$ docker-compose exec kafka kafka-topics.sh --bootstrap-server kafka:9092 --create --topic <YOUR-TOPIC-NAME> --partitions 8 --replication-factor 1
```

### PyFlink

The transaction data will be processed with PyFlink using the Python script [payment_msg_processing.py](payment_msg_proccessing.py).
This script will first map the `provinceId` in the input records to its corresponding province name using a Python UDF, 
and then compute the sum of the transaction amounts for each province. 

### ElasticSearch

ElasticSearch is used to store the upstream processing results and to provide an efficient query service.

### Kibana

Kibana is an open source data visualization dashboard for ElasticSearch. You will use it to visualize 
the results of your PyFlink pipeline.

## Setup

As mentioned, the environment for this walkthrough is based on Docker Compose; It uses a custom image
to spin up Flink (JobManager+TaskManager), Kafka+Zookeeper, the data generator, and Elasticsearch+Kibana
containers.

> Your can find the [docker-compose.yaml](docker-compose.yml) file of the pyflink-walkthrough is located in the `pyflink-walkthrough` root directory.

Reviewer: ```suggestion
Your can find the [docker-compose.yaml](docker-compose.yml) file of the pyflink-walkthrough is located in the `pyflink-walkthrough` root directory.
```
```suggestion
You can find the [docker-compose.yaml](docker-compose.yml) file of the pyflink-walkthrough in the `pyflink-walkthrough` root directory.
```

### Building the Docker image

First, build the Docker image by running:

```bash
docker-compose build
```

### Starting the Playground

Once the Docker image build is complete, run the following command to start the playground:

```bash
docker-compose up -d
```

One way of checking if the playground was successfully started is to access some of the services that are exposed:

1. visiting Flink Web UI [http://localhost:8081](http://localhost:8081).
2. visiting Elasticsearch [http://localhost:9200](http://localhost:9200).
3. visiting Kibana [http://localhost:5601](http://localhost:5601).

**Note:** you may need to wait around 1 minute before all the services come up.

### Stopping the Playground

To stop the playground, run the following command:

```bash
docker-compose down
```

## Running the PyFlink job

1. Submit the PyFlink job.
```shell script
$ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
```

> Navigate to the [Flink Web UI](http://localhost:8081) after the job is submitted successfully, There will be a job in the running job list.

Reviewer: ```suggestion
Navigate to the [Flink Web UI](http://localhost:8081) after the job is submitted successfully. There should be a job in the running job list.
```

> Click the job to turn to the main page of it. You can see that the StreamGraph of the `payment_msg_proccessing` job is consist of two nodes, each of them has parallelism of 1.
> And there is a table in the bottom of the page shows the metrics of every node (bytes received/sent, records received/sent, etc).

Reviewer: ```suggestion
Click the job to get more details. You should see that the `StreamGraph` of the `payment_msg_proccessing` consists of two nodes, each with a parallelism of 1. There is also a table at the bottom of the page that shows some metrics for every node (e.g. bytes received/sent, records received/sent).
```

> ![image](pic/submitted.png)

Reviewer: Did you compile the docs to see how these look? Just asking because if there are no dimensions specified, the images might end up very out of proportion.

Author: Yes, the width of the images would be within the markdown body and automatically scale with page size changes.

Reviewer 2: @morsapaes FWIW, this content isn't targeted for the docs. This is intended to be a github README. I only just realized this myself. Since the other walkthroughs are in the docs, I assume this one was as planned to be as well, but it's not.

Reviewer 2: @shuiqiangchen Do you plan to add a link to this walkthrough from the docs somewhere?

![image](pic/detail.png)    

> 2. Navigate to the [Kibana UI](http://localhost:5601) and choose the pre-created dashboard `payment_dashboard`, There will be a vertical bar chart and a pie chart demonstrating the total amount and the proportion of each province.

Reviewer: ```suggestion
2. Navigate to the [Kibana UI](http://localhost:5601) and choose the pre-created dashboard `payment_dashboard`. There will be a vertical bar chart and a pie chart demonstrating the total amount and the proportion of each province.
```

![image](pic/dash_board.png)

![image](pic/chart.png)

3. Stop the PyFlink job:

Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview) , select the job and click `Cancel` on the upper right side.

![image](pic/cancel.png)

## Further Explorations

If you would like to explore this example more deeply, you can edit [payment_msg_processing.py](payment_msg_proccessing.py)
or create new PyFlink projects that perform more complex processing. You can do this locally under
the `pyflink-walkthrough` directory, since it is mounted on the `jobmanager` docker container.

Ideas:
* Add your own Kafka source table;

> * Create new index for the Elasticsearch sink;

Reviewer: ```suggestion
* Create a new index for the Elasticsearch sink;
```

> * Count the number of transactions, grouped by a 1 minute tumbling windows and payPlatform.

Reviewer: ```suggestion
* Count the number of transactions, grouped by a 1 minute [tumbling window](https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/tableApi.html#tumble-tumbling-windows) and `payPlatform`.
```

After making a modification, you can submit the new job by executing the same command mentioned at 
[Running the PyFlink Job](#running-the-pyflink-job)
```shell script
$ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
```

> Furthermore, you can also [create new kibana dashboards](https://www.elastic.co/guide/en/kibana/7.8/dashboard-create-new-dashboard.html) 

Reviewer: ```suggestion
Furthermore, you can also [create new Kibana dashboards](https://www.elastic.co/guide/en/kibana/7.8/dashboard-create-new-dashboard.html) 
```

3. Stop the PyFlink job:

Visit the Flink Web UI at [http://localhost:8081/#/overview](http://localhost:8081/#/overview) , select the job and click `Cancel` on the upper right side.

![image](pic/cancel.png)

> ## Extension
> 
> You are able to edit the [payment_msg_processing.py](payment_msg_proccessing.py) or create new PyFlink 
> projects to perform more complex processing logic locally on your operating system under the `pyflink-walkthrough` 
> directory since it is mounted on the `jobmanager` docker container. Such as:
> * Creating a new Kafka source table;
> * Creating a new index for the Elasticsearch sink;
> * Calculating the amount of transactions grouped by a 1 minute tumble window and payPlatforms.
> 
> After the modification, you can submit the new job by executing the same command mentioned at 
> [Running the PyFlink Job](#running-the-pyflink-job)
> ```shell script
> $ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
> ```
> 
> Furthermore, you can also [create new kibana dashboards](https://www.elastic.co/guide/en/kibana/7.8/dashboard-create-new-dashboard.html) 
> to visualize more charts of various dimension based on the persistent indexes in Elasticsearch.

Reviewer 2: ```suggestion
## Further Explorations

If you would like to explore this example more deeply, you can edit [payment_msg_processing.py](payment_msg_proccessing.py) or create new PyFlink 
projects that perform more complex processing. You can do this locally under the `pyflink-walkthrough` 
directory, since it is mounted on the `jobmanager` docker container. 

Ideas:
* Add your own Kafka source table;
* Create a new index for the Elasticsearch sink;
* Count the number of transactions, grouped by a 1 minute tumbling windows, and by payPlatform.

After making a modification, you can submit the new job by executing the same command mentioned at 
[Running the PyFlink Job](#running-the-pyflink-job)
```bash
$ docker-compose exec jobmanager ./bin/flink run -py /opt/pyflink-walkthrough/payment_msg_proccessing.py -d
```

Furthermore, you can also [create new kibana dashboards](https://www.elastic.co/guide/en/kibana/7.8/dashboard-create-new-dashboard.html) 
that visualize other aspects of the data in the Elasticsearch.
```
